exchanges/gateio/gateio_websocket_request_spot.go

```diff
@@ -0,0 +1,278 @@
+package gateio
+
+import (
+	"context"
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/order"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/request"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/stream"
+)
+
+var (
+	errBatchSliceEmpty  = errors.New("batch cannot be empty")
+	errNoOrdersToCancel = errors.New("no orders to cancel")
+	errEdgeCaseIssue    = errors.New("edge case issue")
+	errChannelEmpty     = errors.New("channel cannot be empty")
+)
+
+// websocketLogin authenticates the websocket connection
+func (g *Gateio) websocketLogin(ctx context.Context, conn stream.Connection, channel string) (*WebsocketLoginResponse, error) {
+	if conn == nil {
+		return nil, fmt.Errorf("%w: %T", common.ErrNilPointer, conn)
+	}
+
+	if channel == "" {
+		return nil, errChannelEmpty
+	}
+
+	creds, err := g.GetCredentials(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	tn := time.Now().Unix()
+	msg := "api\n" + channel + "\n" + "\n" + strconv.FormatInt(tn, 10)
+	mac := hmac.New(sha512.New, []byte(creds.Secret))
+	if _, err = mac.Write([]byte(msg)); err != nil {
+		return nil, err
+	}
+	signature := hex.EncodeToString(mac.Sum(nil))
+
+	payload := WebsocketPayload{
+		RequestID: strconv.FormatInt(conn.GenerateMessageID(false), 10),
+		APIKey:    creds.Key,
+		Signature: signature,
+		Timestamp: strconv.FormatInt(tn, 10),
+	}
+
+	req := WebsocketRequest{Time: tn, Channel: channel, Event: "api", Payload: payload}
+
+	resp, err := conn.SendMessageReturnResponse(ctx, request.Unset, req.Payload.RequestID, req)
+	if err != nil {
+		return nil, err
+	}
+
+	var inbound WebsocketAPIResponse
+	if err := json.Unmarshal(resp, &inbound); err != nil {
+		return nil, err
+	}
+
+	if inbound.Header.Status != "200" {
+		var wsErr WebsocketErrors
+		if err := json.Unmarshal(inbound.Data, &wsErr.Errors); err != nil {
+			return nil, err
+		}
+		return nil, fmt.Errorf("%s: %s", wsErr.Errors.Label, wsErr.Errors.Message)
+	}
+
+	var result WebsocketLoginResponse
+	return &result, json.Unmarshal(inbound.Data, &result)
+}
+
+// WebsocketOrderPlaceSpot places an order via the websocket connection. You can
+// send multiple orders in a single request. But only for one asset route.
```

Comment:
Yeah 😆 Its more of me trying to map to the exchanges convention [example](https://www.gate.io/docs/developers/apiv4/ws/en/#order-place) and trying to tie it in then on the interface side have it more of a better convention. So in the event of trying to find the relation from their API docs Where is `OrderPlace` for example then you will find a few hits across some files in this package. So prefixed with websocket and tacked Spot on the end, this will also be different for futures trading functions in another PR. That was my reasoning not necessarily correct reasoning I might add. 😆 If you still want me to change I can.

Reply:
Yeah. Okay. I'm still going to vote for standardising more towards our naming, even accepting that this is the "client" side of the package. I'm still hoping to break that line a little more at some point so our client interface package and our GCT adapter package can be separate.

Explicitly: I'd still vote against introducing `Place` here, and use `Submit`.
Reasoning: GCT users who wanted to find this function are GCT centric, not GateIO centric. They're not looking at GateIO specs trying to work out what they can do with GCT, they're looking at GCT docs.

Not a change request. Still just a suggestion.

Reply:
Note: Resolve if you don't want to address :)

Reply:
I won't resolve just yet for others to check this out. @thrasher- @gloriousCode

Reply:
`WebsocketSpotOrderPlace` I think most cleanly matches our standard of naming API functions after their entry in the API docs. If this were a wrapper function I'd want `Submit`

I also think I complained about the function name above without realising that it's the API doc's fault

Reply:
Standardisation of client side methods is great and the desired long term direction, though for now that's to be expected in the wrappers. What I want to see first is integration of all exchange API endpoints and then naming standards can be enforced across all exchanges after we've achieved that since not all exchanges support the same features/endpoints etc. We currently have many PlaceOrder methods across several exchanges which would all need updating as well.

For now, I'm fine with `Websocket(Place/Submit)SpotOrders` or WebsocketSpot(Place/Submit)Orders. I think OrderPlace is a bit funny, despite it being correlated to the API docs

Reply:
[Names have been changed](https://github.com/thrasher-corp/gocryptotrader/pull/1598/commits/544fbccdedbe9b76a80cde4383e3edfc1e184e49)